test/e2e/apps/deployment.go

```diff
@@ -848,45 +848,63 @@ func testProportionalScalingDeployment(f *framework.Framework) {
+	firstRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(firstRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), firstRS)
 	Expect(err).NotTo(HaveOccurred())
 
-	// First rollout's replicaset should have .spec.replicas = 10 - 2 = 8 replicas.
-	newSpecReplicas := replicas - int32(maxUnavailable)
-	framework.Logf("Verifying that first rollout's replicaset has .spec.replicas = %d", newSpecReplicas)
-	Expect(*firstRS.Spec.Replicas).Should(Equal(newSpecReplicas))
+	// Start testing against second rollout's replicaset.
+	secondRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
 
 	maxSurge, err := intstr.GetValueFromIntOrPercent(deployment.Spec.Strategy.RollingUpdate.MaxSurge, int(*(deployment.Spec.Replicas)), false)
 	Expect(err).NotTo(HaveOccurred())
 
-	// Second rollout's replicaset should have .spec.replicas = 10 + 3 - 8 = 5 replicas.
-	newSpecReplicas = replicas + int32(maxSurge) - newSpecReplicas
-	framework.Logf("Verifying that second rollout's replicaset has .spec.replicas = %d", newSpecReplicas)
-	Expect(*secondRS.Spec.Replicas).Should(Equal(int32(newSpecReplicas)))
+	// Second rollout's replicaset should have 0 available replicas.
+	framework.Logf("Waiting for the second rollout's replicaset to have .status.availableReplicas = 0")
+	Expect(framework.WaitForReplicaSetTargetAvailableReplicas(c, secondRS, 0)).NotTo(HaveOccurred())
+
+	// Second rollout's replicaset should have Deployment's (replicas + maxSurge - first RS's replicas) = 10 + 3 - 8 = 5 for .spec.replicas.
+	newReplicas = replicas + int32(maxSurge) - newReplicas
+	framework.Logf("Waiting for the second rollout's replicaset to have .spec.replicas = %d", newReplicas)
+	Expect(framework.WaitForReplicaSetTargetSpecReplicas(c, secondRS, newReplicas)).NotTo(HaveOccurred())
+
+	// The desired replicas wait makes sure that the RS controller has created expected number of pods.
+	framework.Logf("Waiting for the second rollout's replicaset of deployment %q to have desired number of replicas", deploymentName)
+	secondRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(secondRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), secondRS)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Check the deployment's minimum availability.
+	framework.Logf("Verifying that deployment %q has minimum required number of available replicas", deploymentName)
+	if deployment.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", deployment.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
 
-	newReplicas := int32(30)
+	// Scale the deployment to 30 replicas.
+	newReplicas = int32(30)
```

Comment:
@janetkuo I haven't tested for `availableReplicas` after scaling. I will do so once changes for before scaling part look good to you.

Reply:
It's fine to leave it. Scaling is not a rolling update, so the rolling update strategy is likely to be violated.

Reply:
great! The whole test is ready for review. :)

Reply:
Side note: Only Deployment's `replicas` is scaled, but not `maxSurge` or `maxUnavailable`. Therefore, scaling its ReplicaSets proportionally is very likely to violate the strategy.